doc-site/docs/architecture/distributed_sequencer.md

```diff
@@ -0,0 +1,1437 @@
+    D@{ shape: fr-rect, label: "Select<br/>available<br/>coordinator" }
+    E@{ shape: diam, label: "this Is<br/>Coordinator" }
+    I@{ shape: lean-r, label: "Reject<br/>(wrong coordinator)" }
+    G@{ shape: fr-rect, label: "<a href="#subprocess-coordinate-transaction">coordinate transaction</a>" }
+    J@{ shape: fr-circ, label: " " }
+
+
+    A-->B
+    B --> |Behind| C
+    B --> |Ahead| C
+    B --> |Same| D
+    D --> E
+    E --> |Yes| G
+    E --> |No| I
+    G --> J
+```
+
+##### <a name="message-transaction-delegation-rejected"></a>Transaction delegation rejected
+
+The handling of a delegation rejected message depends on the reason for rejection.
+ - if the reason is `MismatchedBlockHeight` and the target delegate is ahead then a new delegation request is sent once the sender has reached a compatible block height.  Compatible block height is defined as a block height in the same block range. 
+ - if the reason is `MismatchedBlockHeight` and the target delegate is behind then a new delegation request is sent on a periodic interval until it is accepted or rejected with a different reason
+ - if the reason is `NotPreferredCoordinator` the `SendTransaction` process is reset
+
+```proto
+message DelegationRequestRejected {
+    string transaction_id = 1;
+    string delegate_node_id = 2;
+    string delegation_id = 3;
+    string contract_address = 4;
+    DelegationRequestRejectedReason reject_reason = 5;
+    optional MismatchedBlockHeightDetail mismatchedBlockHeightDetail = 6; //included if reject_reason is "MismatchedBlockHeight"
+    optional NotPreferredCoordinatorDetail notPreferredCoordinatorDetail = 7; //included if reject_reason is "NotPreferredCoordinator"
+}
+
+enum DelegationRequestRejectedReason{
+  MismatchedBlockHeight = 0;
+  NotPreferredCoordinator = 1;
+}
+
+message MismatchedBlockHeightDetail {
+  int64 provided_block_height = 1;
+  int64 observed_block_height = 2;
+}
+
+message NotPreferredCoordinatorDetail {
+  string preferred_coordinator = 1;
+  google.protobuf.Timestamp latest_heartbeat_time = 2;
+  string latest_heartbeat_id = 3;
+}
+
+```
+
+##### <a name="message-transaction-delegation-accepted"></a>Transaction delegation accepted
+
+If a node receives a `DelegationAccepted` message then it should start to monitor the continued acceptance of that delegation.  It can expect to receive `HeartbeatNotification` messages from the delegate node and for those messages to include the id of the delegated transaction. The `sender` node cannot assume that the `coordinator` node will persist the delegation request.  If the heartbeat messages stop or if the received heartbeat messages do not contain the expected transaction ids, then the sender should retrigger the `HandleTransaction` process to cause the transaction to be re-delegated either to the same delegate, or new delegate or to be coordinated by the sender node itself. Whichever is appropriate for the current point in time.  
```

Comment:
Digressing to an implementation point here... how do we think code could reasonably be constructed for this?

It seems like a sender node will have:
- A database with zero-or-more (maybe 1million) transactions it wishes to get submitted
- Each transaction will be on the same of different contract address (maybe 1million contract addresses)
- Each contract address has a different coordinator committee
- Each transaction might be in one of a number of states (a state machine diagram would be awesome for this)
    - Initial state - candidate to be delegated to someone
    - Delegated to self - I'm the coordinator so I don't need to healthcheck myself
    - Delegated to remote, but not yet requested submission
    - Delegated to remote, submission requested, but submission not occurred
    - Delegated to remote, submission requested, submission approved, but not yet confirmed
    - Delegated to remote, submission confirmed
         - While in any of these states, we might detect a the block range changing means this is the "wrong" coordinator
    - Done. No need to continue tracking
- The decision on whether an action should be performed for any transaction in this DB could be triggered by:
    - A message from the active coordinator for that txn (we have specs for these starting to finalize in this doc)
    - The _lack of_ a message from the active coordinator fro that txn (the subject of this section)
    - The arrival of a new confirmed block
    - The arrival of an on-chain event (from a confirmed block) confirming a transaction
    - The completion of an earlier transaction that directly (through dependency) or indirectly (through memory-management of "in-fight" transactions for that contract)  unlocks processing

Reply:
In our code in Paladin we've chosen to have evaluation loops against the DB and in-memory state, working at various levels.  I guess I'm writing this question, because I'm not 100% sure how to correlate those code modules to the state machine(s) implied by this specification that I can see how it fully scales to "mempool management" of a large sea of transactions with efficient DB query and in-memory management.

Potentially this whole question should be moved out of this issue, into a separate one that is about updating our code to be an implementation of the spec.